src/protojson/WorkerCzarComIssue.h

```diff
@@ -0,0 +1,131 @@
+
+// qserv headers
+#include "protojson/WorkerQueryStatusData.h"
+
+// This header declarations
+namespace lsst::qserv::protojson {
+
+/// This class is used to send/receive a message from the worker to a specific
+/// czar when there has been a communication issue with the worker sending UberJob
+/// file ready messages. If there have been timeouts, the worker will send this
+/// message to the czar immediately after the worker receives a
+/// WorkerQueryStatusData message from the czar (indicating that communication
+/// is now possible).
+/// If communication with the czar has failed for a long time, the worker
+/// will set "_thoughtCzarWasDead" and delete all incomplete work associated
+/// with that czar. Result files will remain until garbage cleanup or the czar
+/// calls for their removal.
+/// TODO:UJ  UberJob complete messages that failed to be sent to the czar
+/// TODO:UJ  will be added to this message. (uberjob file response)
+/// Upon successful completion, the worker will clear all values set by the
+/// the czar.
+/// Currently, this message is expected to only be needed rarely.
+class WorkerCzarComIssue {
+public:
+    using Ptr = std::shared_ptr<WorkerCzarComIssue>;
+
+    WorkerCzarComIssue() = delete;
+    ~WorkerCzarComIssue() = default;
+
+    std::string cName(const char* funcN) { return std::string("WorkerCzarComIssue") + funcN; }
+
+    static Ptr create(std::string const& replicationInstanceId_, std::string const& replicationAuthKey_) {
+        return Ptr(new WorkerCzarComIssue(replicationInstanceId_, replicationAuthKey_));
+    }
+
+    static Ptr createFromJson(nlohmann::json const& workerJson, std::string const& replicationInstanceId_,
+                              std::string const& replicationAuthKey_);
+
+    void setThoughtCzarWasDead(bool wasDead) {
+        std::lock_guard lg(_wciMtx);
+        _thoughtCzarWasDead = wasDead;
+    }
+
+    bool getThoughtCzarWasDead() const { return _thoughtCzarWasDead; }
+
+    /// Return true if there is a reason this WorkerCzarComIssue should be sent to this czar.
+    bool needToSend() const {
+        std::lock_guard lg(_wciMtx);
+        // TODO:UJ or list of failed transmits not empty.
+        return _thoughtCzarWasDead;
+    }
+
+    /// Set the contact information for the appropriate czar and worker.
+    void setContactInfo(WorkerContactInfo::Ptr const& wInfo_, CzarContactInfo::Ptr const& czInfo_) {
+        std::lock_guard lgWci(_wciMtx);
+        if (_wInfo == nullptr && wInfo_ != nullptr) _wInfo = wInfo_;
+        if (_czInfo == nullptr && czInfo_ != nullptr) _czInfo = czInfo_;
+    }
+
+    CzarContactInfo::Ptr getCzarInfo() const {
+        std::lock_guard lgWci(_wciMtx);
+        return _czInfo;
+    }
+
+    WorkerContactInfo::Ptr getWorkerInfo() const {
+        std::lock_guard lgWci(_wciMtx);
+        return _wInfo;
+    }
+
+    /// Return a json version of the contents of this class.
+    std::shared_ptr<nlohmann::json> serializeJson();
```

Comment:
This has to be:
```
nlohmann::json toJson() const;
```

Reply:
I agree with const and name change. However, it's using a pointer as the json gets passed around as a function argument. It's very easy to make sure that a relatively large object isn't being copied when you're using a pointer to that object.

Reply:
Agreed. Then it makes sense.